### PR TITLE
Support for JSON & Binary Config Files

### PR DESCRIPTION
**Pull request by deprilula28**
## Description

Adds support for JSON & Binary config files, under a feature in amethyst_config

## Additions

- Adds the ConfigFormat enum, with Ron, Json & Binary
- Changes the main example on amethyst_config to read & write from all the supported formats
- Adds the "json" and "binary" features for support on those file types

## Modifications

- Deprecates Config.write(AsRef<Path>) in favor of Config.write_format(ConfigFormat, AsRef<Path>)
- Deprecates Config.load_bytes(&[u8]) in favor of Config.write_format(ConfigFormat, &[u8])

## PR Checklist

By placing an x in the boxes I certify that I have:

- [x] Updated the content of the book if this PR would make the book outdated.
- [x] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [x] Added unit tests for new code added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

If this modified or created any rs files:

- [x] Ran `cargo +stable fmt --all`
- [x] Ran `cargo clippy --all --features "empty"`
- [x] Ran `cargo test --all --features "empty"` (Also with the binary & json features enabled)

## Other Notes
Output for `cargo run --example main --features json,binary` in `amethyst_config`:
```rust
RON Config Result:
ExampleConfig {
    display: DisplayConfig {
        title: "Amethyst game",
        brightness: 1.0,
        fullscreen: false,
        dimensions: (
            1024,
            768,
        ),
        min_dimensions: None,
        max_dimensions: None,
        vsync: true,
        multisampling: 0,
        visibility: true,
    },
    logging: LoggingConfig {
        file_path: "new_project.log",
        output_level: "warn",
        logging_level: "debug",
    },
}
JSON Config Result:
ExampleConfig {
    display: DisplayConfig {
        title: "Amethyst game",
        brightness: 1.0,
        fullscreen: false,
        dimensions: (
            1024,
            768,
        ),
        min_dimensions: None,
        max_dimensions: None,
        vsync: true,
        multisampling: 0,
        visibility: true,
    },
    logging: LoggingConfig {
        file_path: "new_project.log",
        output_level: "warn",
        logging_level: "debug",
    },
}
Binary Config Result:
ExampleConfig {
    display: DisplayConfig {
        title: "Amethyst game",
        brightness: 1.0,
        fullscreen: false,
        dimensions: (
            1024,
            768,
        ),
        min_dimensions: None,
        max_dimensions: None,
        vsync: true,
        multisampling: 0,
        visibility: true,
    },
    logging: LoggingConfig {
        file_path: "new_project.log",
        output_level: "warn",
        logging_level: "debug",
    },
}
```

### CODE REVIEW

**Comment by deprilula28** [comment]
Forgot to mention, this is issue #2387

File: amethyst_config/src/lib.rs [hunk 1]
```diff
@@ -107,10 +152,20 @@ where
...
+    fn write_format<P: AsRef<Path>>(
+        &self,
+        format: ConfigFormat,
+        path: P,
+    ) -> Result<(), ConfigError>;
+    
     /// Writes a configuration structure to a file.
-    fn write<P: AsRef<Path>>(&self, path: P) -> Result<(), ConfigError>;
+    #[deprecated(note = "use `write_format` instead")]
+    fn write<P: AsRef<Path>>(&self, format: ConfigFormat, path: P) -> Result<(), ConfigError> {
```
Comment (chemicstry) [inline amethyst_config/src/lib.rs:166]: API changed, you probably want to drop the `format: ConfigFormat` parameter.

File: amethyst_config/examples/config.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    "vsync": true,
+    "multisampling": 0,
+    "visibility": true
+  },
+  "logging": {
+    "file_path": "new_project.log",
+    "output_level": "warn",
+    "logging_level": "debug"
+  }
+}
```
Comment (chemicstry) [inline amethyst_config/examples/config.json:21]: Newline

File: amethyst_config/src/lib.rs [hunk 3]
```diff
@@ -107,10 +152,20 @@ where
...
+    fn load_bytes_format(format: ConfigFormat, bytes: &[u8]) -> Result<Self, ConfigError>;
+    
+    /// Writes a configuration structure to a file.
+    fn write_format<P: AsRef<Path>>(
+        &self,
+        format: ConfigFormat,
+        path: P,
+    ) -> Result<(), ConfigError>;
+    
     /// Writes a configuration structure to a file.
```
Comment (chemicstry) [inline amethyst_config/src/lib.rs:164]: `/// Writes a configuration structure to a file using Ron format`

**Comment by chemicstry** [comment]
I don't really like all the `_format` function names, but I guess it's the only way to avoid breaking old API.

Thanks for the PR!

File: amethyst_config/src/lib.rs (hunk 1 above)
Comment (deprilula28) [reply amethyst_config/src/lib.rs:166]: That's what I meant to do sorry

File: amethyst_config/examples/config.json (hunk 2 above)
Comment (deprilula28) [reply amethyst_config/examples/config.json:21]: I'm not sure what you mean? This is the output from serde json

Comment (chemicstry) [reply amethyst_config/examples/config.json:21]: It's usual practice for files to end with a newline as most of the diff tools expect this format. Github also reminds you of this if you look at the changes tab.

But if it's how serde outputs then leave it as is

**Comment by CleanCut** [comment]
@chemicstry I have some questions (I'm actually asking, not implying they must be done):

- Should this go in 0.15.1?
- Should there be API documentation for json/bincode formats?
- Should there be a page in the book created/updated about setting up the config with the new formats?

**Comment by CleanCut** [comment]
Related, https://github.com/amethyst/amethyst/issues/2387 specifically requests examples of the same config in different formats.  Perhaps it would be good to put some version the output for `cargo run --example main --features json,binary` into the docs somewhere.

**Comment by chemicstry** [comment]
@CleanCut I think we should release 0.15.1 just before legion is ready for merge, because these minor PRs keep poping up. I guess the no PRs until legion policy is not advertised enough. Throwing away PRs is not an option imo.

Regarding docs, it is not really possible to document format as `amethyst_config` only provides implementation for any struct to be serialized. Currently most examples use it to deserialize `DisplayConfig`, but adding format specific documentation for that would be a maintenance nightmare to keep it in sync with the code.

I see the book as an introduction to amethyst and it does not have to cover all possible use cases. For more in depth knowledge look at either `cargo doc` or examples. Although `cargo doc` doesn't help in this case because formats are feature gated. Maybe adding `README.md` to `amethyst_config` mentioning how to compile example would be enough?

**Comment by deprilula28** [comment]
Oh I'm really sorry I didn't see that policy

**Comment by chemicstry** [comment]
No problem :) It wasn't communicated properly by us

**Comment by CleanCut** [comment]
> I guess the no PRs until legion policy is not advertised enough.

> No problem :) It wasn't communicated properly by us

True..._I_ wasn't aware of that policy. 😊 Seems like the no-more-stop-the-world-changes decision (🎉) and the fact that we've got months of changes since 0.15.0 pretty much override the policy anyway.

>  I think we should release 0.15.1 just before legion is ready for merge, because these minor PRs keep poping up. [snip] Throwing away PRs is not an option imo.

Throwing away PRs is definitely not an option. I'd go even further and say that we ought to do patch releases _frequently_ (non-breaking fixes for 0.15.1 or 0.15.2, etc.).  I'm going to run through the release process with @AlveLarsson for the first time myself for 0.15.1, so I'll learn firsthand how feasible that really is.  If it's not too bad, I'd say do 0.15.1 now and 0.15.2 right before legion.  I'm well-positioned to make the release process better since I'll be both participating in it and [improving CI/CD](https://github.com/amethyst/amethyst/issues/2407).  CI/CD ought to do the bulk of the release work anyway -- that's the whole point of the CD (continuous delivery) part.  My dream is that we pick a release number and CI/CD does all the rest, except things like blog posts and advertisements for major releases.  I think the minor releases could be fully automated, though, down to a blog post that consists of the latest changelog stuff.

**Comment by CleanCut** [comment]
Update: We had a good member+leadership discussion on Discord and some things got clarified.

- We are going to very strict about only accepting minor bug fixes, documentation updates, etc. for the 0.15.x series to enable most active contributors to focus on the legion port.
- This PR is clearly a new feature to the project, not a fix, so I'm marking it for 0.17.0
- Once 0.16.0 has been released, we can remove the "status: blocked" and move forward on this.

**Comment by jojolepro** [comment]
We are now accepting those pull requests again. If it can be updated to the new main branch, please do so. Thanks! <3

**Comment by CleanCut** [comment]
@ezpuzz (since you seem to be the one who is most involved in this).

Since I had a bit of time tonight, I sat down to see if I could help with this PR.

After rebasing and adding an additional fixup commit locally, there seems to be some new `rusty-hook` magic that plays with git hooks, but has a broken config (which I have not touched):

```rust
$ gcam "Minor fixup due to code drift in dependencies."         
[rusty-hook] Upgrading rusty-hook cli...
[rusty-hook] This may take a few seconds...
Invalid rusty-hook config file
[rusty-hook] Configured hook command failed
[rusty-hook] pre-commit hook rejected
```

```
$ cat .rusty-hook.toml 
[hooks]
pre-commit = [
    "cargo check --workspace --all-targets --features empty,optional",
    "cargo fix --workspace --features empty,optional --allow-dirty --allow-staged",
    "cargo +nightly clippy -Z unstable-options --workspace --features empty,optional --fix --allow-dirty --allow-staged",
    "cargo +nightly fmt --all",
]
pre-push = ["cargo test --all-targets --workspace --features empty,optional"]

[logging]
verbose = true
```

**Comment by ezpuzz** [comment]
You'll need rusty-hook master or just use git commit --no-verify.  I kept
pushing broken code and it was frustrating.  Sorry I didn't get to document
it as well as I liked.  Perhaps we can make it optional somehow.

**Comment by ezpuzz** [comment]
Maybe cargo-husky is a better solution.  I'm not seeking to block anyone from committing or pushing their changes.   I will make a PR to remove rusty-hook for now.

**Comment by CleanCut** [comment]
>  I will make a PR to remove rusty-hook for now.

I think that is a good idea, rusty hook breakage also prevented me from using other tools (like GitHub's `gh`) that interact with the repo, and even something like creating a new branch was annoyingly taking 5-10 seconds due to this integration wanting to incessantly self-update.

I would be in favor of an _opt-in_ solution for client-side checking, and leave the mandatory checks to CI.  I'd love to be able to run some `cargo ci`-like command upon demand to run as many of the CI-like checks locally as possible.  My two cents, anyway.

**Comment by CleanCut** [comment]
I rebased and attempted fixing up this branch and pushed it up as a new PR: #2586

**Pull request closed without merging**